dali/operators/reader/loader/webdataset_loader.h

```diff
@@ -0,0 +1,88 @@
+#ifndef DALI_OPERATORS_READER_LOADER_WEBDATASET_LOADER_H_
+#define DALI_OPERATORS_READER_LOADER_WEBDATASET_LOADER_H_
+
+#include <fstream>
+#include <set>
+#include <string>
+#include <unordered_map>
+#include <vector>
+#include "dali/operators/reader/loader/loader.h"
+#include "dali/operators/reader/loader/webdataset/tar_utils.h"
+#include "dali/pipeline/data/tensor.h"
+
+namespace dali {
+namespace {
+
+const char kExtDelim = ';';
+
+}  // namespace
+
+class WebdatasetLoader : public Loader<CPUBackend, vector<Tensor<CPUBackend>>> {
+ public:
+  static const std::set<DALIDataType> kSupportedTypes;
+  explicit WebdatasetLoader(const OpSpec& spec);
+  ~WebdatasetLoader() override;
+
+  void PrepareEmpty(std::vector<Tensor<CPUBackend>>&) override;
+  void ReadSample(std::vector<Tensor<CPUBackend>>&) override;
+
+ protected:
+  Index SizeImpl() override;
+  void PrepareMetadataImpl() override;
+  void Reset(bool wrap_to_shard) override;
+
+  enum class MissingExt
+  {
+    Empty,
+    Skip,
+    Raise,
+    Invalid
+  };
+  static MissingExt Str2MissingExt(std::string);
+
+  std::vector<std::string> uris_;
+  std::vector<std::string> configs_;
+  std::vector<std::set<std::string>> ext_;
+  std::unordered_map<std::string, std::vector<size_t>>
+      ext_map_;  // maps an extension to sample indicies
+  MissingExt missing_component_behavior_;
+  std::vector<DALIDataType> dtype_;
+
+ private:
+  size_t MaxCommonDtypeSize(const std::string& extension) const;
+  void SetDataPointer(std::vector<Tensor<CPUBackend>>& sample, std::vector<bool>& sample_was_set,
+                      const std::string& extension, const std::string& source_info,
+                      std::shared_ptr<void> data, int64_t size) const;
+  uint8_t* ShareDataPointer(std::vector<Tensor<CPUBackend>>& sample,
+                            std::vector<bool>& sample_was_set, const std::string& extension,
+                            const std::string& source_info, int64_t size,
+                            size_t dtype_max_size) const;
+  void MarkCached(std::vector<Tensor<CPUBackend>>& sample, std::vector<bool>& sample_was_set,
+                  const std::string& extension, const std::string& source_info) const;
+
+  struct SampleConfig {
+    int64_t start_offset;
+    int64_t end_offset;
+    std::set<std::string> extensions;
```

Comment:
Can each sample have different set of extensions?

Reply:
yes